src/python/pants/engine/target.py

```diff
@@ -1343,15 +1343,39 @@ def filespec(self) -> Filespec:
         )
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass(unsafe_hash=True)
 class HydrateSourcesRequest:
     field: Sources
+    valid_sources_types: Tuple[Type[Sources], ...]
```

Comment:
Naming wise, the word "valid" would seem to imply that the request will fail if the sources can't be converted to one of these types. "desired" would be a bit verbose. Maybe just `output_` ...?